torchchat/generate.py

```diff
@@ -1246,20 +1557,20 @@ def main(args):
 
         for _ in gen.chat(generator_args):
             pass
-    else:
-        dist_gen = DistributedGenerator(
-            args.model,
-            builder_args,
-            tokenizer_args,
-            generator_args,
-            args.profile,
-            args.quantize,
-            args.draft_quantize,
-        )
 
-        response = ""
-        for output in dist_gen.generate(generator_args.prompt):
-            response += output.text
-
-        print(f"Model output: {response}")
-        dist_gen.shutdown()
+def main(args):
+    builder_args = BuilderArgs.from_args(args)
+    
+    if builder_args.distributed:
+        world_size = builder_args.tp * builder_args.pp
+
+        ctx = mp.get_context('spawn')
+        with futures.ProcessPoolExecutor(max_workers=world_size-1, mp_context=ctx) as executor:
+            for i in range(1,world_size):
+                fn = partial(run_generator, args, i)
+                executor.submit(setup_env, world_size, i, fn)
+            #Starting rank 0
+            fn = partial(run_generator, args, 0)
+            setup_env(world_size, 0, fn)
+    else:
+        run_generator(args)
```

Comment:
Shuold `setup_env` also be triggering the generator?

Maybe it's just a naming change (run_with_setup~?), but setup_env doesn't make me think that the fn is called

Reply:
Yes, valid point. I've renamed it to run_in_dist_env to reflect that it runs the target as well as connecting it to distributed.